server/src/actors/healthcheck.rs

```diff
@@ -0,0 +1,67 @@
+use ::actix::prelude::*;
+
+use futures::future;
+use futures::prelude::*;
+
+use crate::actors::controller::{Controller, Shutdown, Subscribe, TimeoutError};
+use crate::actors::upstream::{IsAuthenticated, UpstreamRelay};
+
+pub struct Healthcheck {
+    is_shutting_down: bool,
+    upstream: Addr<UpstreamRelay>,
+}
+
+impl Healthcheck {
+    pub fn new(upstream: Addr<UpstreamRelay>) -> Self {
+        Healthcheck {
+            is_shutting_down: false,
+            upstream,
+        }
+    }
+}
+
+impl Actor for Healthcheck {
+    type Context = Context<Self>;
+
+    fn started(&mut self, context: &mut Self::Context) {
+        Controller::from_registry().do_send(Subscribe(context.address().recipient()));
+    }
+}
+
+impl Handler<Shutdown> for Healthcheck {
+    type Result = Result<(), TimeoutError>;
+
+    fn handle(&mut self, _message: Shutdown, _context: &mut Self::Context) -> Self::Result {
+        self.is_shutting_down = true;
+        Ok(())
+    }
+}
+
+pub enum IsHealthy {
+    /// Check if the Relay is alive at all.
+    Liveness,
+    /// Check if the Relay is in a state where the load balancer should route traffic to it (i.e.
+    /// it's both live/alive and not too busy).
+    Readiness,
+}
+
+impl Message for IsHealthy {
+    type Result = Result<bool, ()>;
+}
+
+impl Handler<IsHealthy> for Healthcheck {
+    type Result = ResponseFuture<bool, ()>;
```

Comment:
You could directly return `bool` here, which implements `MessageResponse` iirc. Just `unwrap_or_default` the request to upstream.

Reply:
Doesn't work because `send()` returns a `Future<Item = Result<bool, ???>, Error = MailboxError>` which I need to poll.

Reply:
Excellent point! 😄